File: scheduler/core/src/main/java/io/orkes/conductor/scheduler/model/WorkflowSchedule.java
package io.orkes.conductor.scheduler.model;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import com.netflix.conductor.common.metadata.workflow.StartWorkflowRequest;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.ToString;

@Getter
@Setter
@ToString
@NoArgsConstructor
@AllArgsConstructor
@EqualsAndHashCode
@Builder
public class WorkflowSchedule {

    private String name;
    private String cronExpression;
    private boolean runCatchupScheduleInstances;
    private boolean paused;
    private String pausedReason;
    @EqualsAndHashCode.Exclude private StartWorkflowRequest startWorkflowRequest;
    @Builder.Default private String zoneId = "UTC";

    /**
     * List of cron schedules for multi-expression configuration. When this list is not empty, it
     * takes priority over the single cronExpression/zoneId fields. Each entry in the list can have
     * its own cron expression and timezone. If a timezone is not specified for an entry, UTC is
     * used as the default.
     */
    @Setter(lombok.AccessLevel.NONE)
    @Getter(lombok.AccessLevel.NONE)
    private List<CronSchedule> cronSchedules;

    @JsonProperty("cronSchedules")
    public List<CronSchedule> getCronSchedules() {
        return cronSchedules != null ? cronSchedules : new ArrayList<>();
    }

    @JsonProperty("cronSchedules")
    public void setCronSchedules(List<CronSchedule> cronSchedules) {
        if (cronSchedules == null) {
            this.cronSchedules = new ArrayList<>();
        } else {
            // mutable
            this.cronSchedules = new ArrayList<>(cronSchedules);
        }
    }

    private Long scheduleStartTime;
    private Long scheduleEndTime;

    private Long createTime;
    private Long updatedTime;
    private String createdBy;
    private String updatedBy;
    private String description;
    private Long nextRunTime;

    /**
     * Returns the effective list of cron schedules. If cronSchedules list is not empty, returns
     * that list. Otherwise, returns a single-element list with the cronExpression and zoneId from
     * this schedule.
     *
     * @return list of effective cron schedules, never null
     */
    @JsonIgnore
    public List<CronSchedule> getEffectiveCronSchedules() {
        if (cronSchedules != null && !cronSchedules.isEmpty()) {
            return cronSchedules;
        }
        if (cronExpression != null) {
            return Collections.singletonList(
                    CronSchedule.builder()
                            .cronExpression(cronExpression)
                            .zoneId(zoneId != null ? zoneId : "UTC")
                            .build());
        }
        return Collections.emptyList();
    }

    @JsonIgnore
    public boolean hasMultipleCronSchedules() {
        return cronSchedules != null && !cronSchedules.isEmpty();
    }
}

File: scheduler/core/src/main/typescript/io/orkes/conductor/scheduler/model/WorkflowSchedule.ts
import { CronSchedule } from './CronSchedule';

export class WorkflowSchedule {
    name?: string;
    cronExpression?: string;
    runCatchupScheduleInstances?: boolean;
    paused?: boolean;
    pausedReason?: string;
    startWorkflowRequest?: any;
    zoneId: string = 'UTC';
    private _cronSchedules: CronSchedule[] = [];

    get cronSchedules(): CronSchedule[] {
        return this._cronSchedules;
    }

    set cronSchedules(cronSchedules: CronSchedule[]) {
        if (!cronSchedules) {
            this._cronSchedules = [];
        } else {
            this._cronSchedules = [...cronSchedules];
        }
    }

    scheduleStartTime?: number;
    scheduleEndTime?: number;

    createTime?: number;
    updatedTime?: number;
    createdBy?: string;
    updatedBy?: string;
    description?: string;
    nextRunTime?: number;

    getEffectiveCronSchedules(): CronSchedule[] {
        if (this._cronSchedules && this._cronSchedules.length > 0) {
            return this._cronSchedules;
        }
        if (this.cronExpression) {
            const schedule = new CronSchedule();
            schedule.cronExpression = this.cronExpression;
            schedule.zoneId = this.zoneId != null ? this.zoneId : 'UTC';
            return [schedule];
        }
        return [];
    }

    hasMultipleCronSchedules(): boolean {
        return this._cronSchedules != null && this._cronSchedules.length > 0;
    }
}
